backend/schemamirror/service.go

```diff
@@ -0,0 +1,229 @@
+package schemamirror
+
+import (
+	"context"
+	"maps"
+	"slices"
+	"sync/atomic"
+	"time"
+
+	"connectrpc.com/connect"
+	"github.com/alecthomas/errors"
+
+	ftlv1 "github.com/block/ftl/backend/protos/xyz/block/ftl/v1"
+	"github.com/block/ftl/backend/protos/xyz/block/ftl/v1/ftlv1connect"
+	"github.com/block/ftl/common/key"
+	"github.com/block/ftl/common/log"
+	schemapb "github.com/block/ftl/common/protos/xyz/block/ftl/schema/v1"
+	"github.com/block/ftl/common/schema"
+	islices "github.com/block/ftl/common/slices"
+	"github.com/block/ftl/internal/channels"
+	"github.com/block/ftl/internal/schema/schemaeventsource"
+)
+
+type Service struct {
+	// receiving *sync.Mutex
+	receiving   *atomic.Bool
+	eventSource *schemaeventsource.EventSource
+}
+
+func New(ctx context.Context) *Service {
+	svc := &Service{
+		// receiving:   &sync.Mutex{},
+		receiving:   &atomic.Bool{},
+		eventSource: schemaeventsource.NewUnattached(),
+	}
+	logger := log.FromContext(ctx)
+	go func() {
+		var wasReceiving bool
+		for {
+			select {
+			case <-time.After(time.Second * 30):
+				isReceiving := svc.receiving.Load()
+				// Log an error if atleast 2 consecutive checks show that we are not receiving schema changes.
+				if !wasReceiving && !isReceiving {
+					logger.Logf(log.Error, "Schema mirror does not have a push stream connected")
+				}
+				wasReceiving = isReceiving
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
```

Comment:
This seems like it could be a method rather than a lambda?

Reply:
yeh good point